packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -5076,21 +5076,79 @@ class _ScribblePlaceholder extends WidgetSpan {
   }
 }
 
-/// A text boundary that uses code units as logical boundaries.
+/// A text boundary that uses code points as logical boundaries.
 ///
-/// This text boundary treats every character in input string as an utf-16 code
-/// unit. This can be useful when handling text without any grapheme cluster,
-/// e.g. password input in [EditableText]. If you are handling text that may
-/// include grapheme clusters, consider using [CharacterBoundary].
-class _CodeUnitBoundary extends TextBoundary {
-  const _CodeUnitBoundary(this._text);
+/// A code point represents a single character. This may be smaller than what is
+/// represented by a user-perceived character, or grapheme. For example, a
+/// single grapheme (in this case a Unicode extended grapheme cluster) like
+/// "👨‍👩‍👦" consists of five code points: the man emoji, a zero
+/// width joiner, the woman emoji, another zero width joiner, and the boy emoji.
+/// The [String] has a length of eight because each emoji consists of two code
+/// units.
+///
+/// Code units are the units by which Dart's String class is measured, which is
+/// encoded in UTF-16.
+///
+/// See also:
+///
+///  * [String.runes], which deals with code points like this class.
+///  * [String.characters], which deals with graphemes.
+///  * [CharacterBoundary], which is a [TextBoundary] like this class, but whose
+///    boundaries are graphemes instead of code points.
+class _CodePointBoundary extends TextBoundary {
+  const _CodePointBoundary(this._text);
 
   final String _text;
 
+  // Returns true if the given position falls in the middle of a single code
+  // point in _text.
+  bool _breaksCodePoint(int position) {
+    assert(position > 0 && position < _text.length && _text.length > 1);
+    // A string containing the two code units that surround the given position.
+    final String surroundingCodeUnits = _text.substring(
+      position - 1,
+      position + 1,
+    );
+
+    // If the surrounding code units are a single code point, then the position
+    // was in the center of that code point.
+    // a[a|b]
+    // [a|b]b
+    // [a|b]c
+    // [a|a]b
+    // a[b|b]
+    return surroundingCodeUnits.runes.length == 1;
+  }
+
   @override
-  int getLeadingTextBoundaryAt(int position) => position.clamp(0, _text.length); // ignore_clamp_double_lint
+  int? getLeadingTextBoundaryAt(int position) {
+    if (position > _text.length || position < 0) {
```

Comment:
If position == _text.length looks like this is going to throw.

Reply:
The documentation promises to return the closest index <= `position` that points to the start of a code point. So when position > _text.length it is supposed to return `getLeadingAt(_text.length - 1)`.

Reply:
Good catch.  I think I've cleaned up these edge cases in [b4c9c47](https://github.com/flutter/flutter/pull/123366/commits/b4c9c4758fcd70d037c5a1646f99595e56da2fdc) now.